sdk/core/core-lro/src/createPoller.ts

```diff
@@ -0,0 +1,162 @@
+  setRunning: (state) => (state.status = "running"),
+  setSucceeded: (state) => (state.status = "succeeded"),
+  setFailed: (state) => (state.status = "failed"),
+
+  getError: (state) => state.error,
+  getResult: (state) => state.result,
+  isCanceled: (state) => state.status === "canceled",
+  isCanceling: (state) => state.status === "canceling",
+  isFailed: (state) => state.status === "failed",
+  isRunning: (state) => state.status === "running",
+  isSucceeded: (state) => state.status === "succeeded",
+});
+
+/**
+ * Creates a poller that can be used to poll a long-running operation.
+ * @param lro - Description of the long-running operation
+ * @param options - options to configure the poller
+ * @returns an initialized poller
+ */
+export async function createPoller<TResult, TState extends OperationState<TResult>>(
+  lro: LongRunningOperation,
+  options?: CreatePollerOptions<TResult, TState>
+): Promise<SimplePollerLike<TState, TResult>> {
+  const {
+    resourceLocationConfig,
+    processResult,
+    updateState,
+    withPollingUrl: withPollingUrlCallback,
+    intervalInMs = POLL_INTERVAL_IN_MS,
+    restoreFrom,
+  } = options || {};
+  const { requestMethod, requestPath } = lro;
+  const stateProxy = createStateProxy<TResult, TState>();
+  const withPollingUrl = withPollingUrlCallback
+    ? (() => {
+        let called = false;
+        return (pollingUrl: string, isUpdated: boolean) => {
+          if (isUpdated) withPollingUrlCallback(pollingUrl);
+          else if (!called) withPollingUrlCallback(pollingUrl);
+          called = true;
+        };
+      })()
+    : undefined;
+  const state: RestorableOperationState<TState> = restoreFrom
+    ? deserializeState(restoreFrom)
+    : await initOperation({
+        lro,
+        stateProxy,
+        requestPath,
+        requestMethod,
+        resourceLocationConfig,
+        processResult,
+        withPollingUrl,
+      });
+  let resultPromise: Promise<TResult> | undefined;
+  let cancelJob: (() => void) | undefined;
+  // Progress handlers
+  type Handler = (state: TState) => void;
+  const handlers = new Map<symbol, Handler>();
+  const handleProgressEvents = async (): Promise<void> => handlers.forEach((h) => h(state));
+
+  let currentPollIntervalInMs = intervalInMs;
+
+  const poller: SimplePollerLike<TState, TResult> = {
+    getOperationState: () => state,
+    getResult: () => state.result,
+    isDone: () => ["succeeded", "failed", "canceled"].includes(state.status),
+    isStopped: () => resultPromise === undefined,
+    stopPolling: () => cancelJob?.(),
+    toString: () =>
+      JSON.stringify({
+        state,
+      }),
+    onProgress: (callback: (state: TState) => void) => {
+      const s = Symbol();
+      handlers.set(s, callback);
+      return () => handlers.delete(s);
+    },
+    pollUntilDone: () =>
+      (resultPromise ??= (async () => {
+        if (!poller.isDone()) {
+          await poller.poll();
```

Comment:
This is an annoying thing I just thought about, but I think we might need to create an abort controller and pass its signal along to the `poll` calls so that if the operation is stopped we can abort the poll call. I'm not sure: what does `Poller` do in this case?

Reply:
`PollerLike.pollUntilDone` doesn't accept an abort controller as input but I agree it probably should.

Reply:
I added support for it in 69d3091.

Reply:
wait, I think I misinterpreted your comment, it seems like you want to stop the poll call if the operation is stopped but I am not sure what do you mean exactly? Like, if the response to one poll call says the operation is in a terminal state, the next poll call should be aborted? why would there be a next poll call?

Reply:
I'm suggesting to instantiate an abort controller in the worker promise and pass it to `poll` ourselves, so that we can abort `poll` if a user calls `stopPolling`, though I agree we probably _should_ accept an abort signal as an option to pollUntilDone

Reply:
Ah you mean the poller stopping and not the operation itself. `Poller` doesn't do anything in this regard. I will think about this scenario will probably address it in another PR since this one is already getting too big.

Reply:
Ok I addressed both points in [b240fb1](https://github.com/Azure/azure-sdk-for-js/pull/22753/commits/b240fb11a268f8ac8a680e9ee3c3f2628fed8fad).